Linking.md

```diff
@@ -165,10 +166,34 @@ where a `syminfo` is encoded as:
 
 | Field        | Type           | Description                                 |
 | -------------| -------------- | ------------------------------------------- |
-| name_len     | `varuint32`    | length of `name_str` in bytes               |
-| name_str     | `bytes`        | UTF-8 encoding of the name                  |
+| kind         | `varuint32`    | The symbol type.  One of:                   |
+|              |                |   `0` - SYMTAB_FUNCTION                     |
+|              |                |   `1` - SYMTAB_DATA                         |
+|              |                |   `2` - SYMTAB_GLOBAL                       |
 | flags        | `varuint32`    | a bitfield containing flags for this symbol |
 
+For functions and globals, we reference an existing import/export which contains
+the name, whether the symbol is defined, its type, and so on:
+
+| Field        | Type           | Description                                 |
+| -------------| -------------- | ------------------------------------------- |
+| wasm_index   | `varuint32`    | Specifies an import or                      |
```

Comment:
I would just call this `index`

I also just realized a possible flaw in the symbol table: We don't have nice way address exports.  There could be multiple exports of a given function and there is no such things as an export index space, so there is no way to address them.

What we are trying to do here is avoid duplicating the string names for functions... so perhaps we only do that for imports.

How about this:
```
| name_type   | `varuing32`    | Can be one of NAME_INLINE, NAME_FUNC_IMPORT  

If name_type is NAME_INLINE then the symbol name is specified in the symbol table as a string, otherwise a function index is specified that must point to a function import (which in turn must have a unique name).
```        
       
Then we don't both to export functions as wasm exports.

Basicaly I'm suggesting that we move even further from using imports/exports and only used them for function imports.

Reply:
Perhaps name_type could be a bit in the symbol flags and then the name specifier could be common to all symbols  i.e. based on the symbols flags.

Reply:
We do have a way to address exports: exactly as I've described in the text here! In the table I've written out the definition of the index space, indexing the imports followed by the exports (so it is a new index space, but it's what we need). We can use exactly one import or export per symbol, there's no problem with having multiple exports for a function, as long as we index into the array of imports+exports.

If you don't want an export for some symbols - then we're back to where we were before, and we'll require extra fields in the symtab to store the information that would be in the export (namely, the function index).

Actually there's a good reason we wouldn't want to export some symbols. We could make it so that the internal linkage symbols aren't exported, and external linkage are. (NB internal/local linkage can't generate in import). The motivation would be to allow internal linkage functions to have colliding names. Basically getting rid of the "BudgeName" fix I put in.

How about this for a reasonable modification: we allow the import/export specifier to be optional - in which case it's forced to be an export (not an import). Hence all undefined symbols (which must be external linkage) have an import, and defined symbols have an export depending on internal/external linkage.

In that case, we'd need the following fields for FUNCTION/GLOBAL symbols:
* Kind and Flags
* ImportExportIndex (or -1 to mean "has no import/export")
* optional Name (only written out if ImportExportIndex is -1)
* optional ObjectIndex (only written out if ImportExportIndex is -1), signifies the Function/Global that's being exported. This is the index that _would_ be in the export if it existed

The names in the symbol table are only unique across external linkage symbols (of all types)

Reply:
I really don't want to invent a new import or export index space.  I feel pretty strongly that if the symbol table is going to reference any index spaces it should be referencing existing ones (i.e. the ObjectIndex above makes total sense to me).

I think I prefer something like this:

- Kind
- Flags
- IsNamedByImport (boolean)
- Optional Name (only specified for not IsNamedByImport
- Optional Index (the function or global index if IsNamedByImport)

In this model there are not exports, and the only things that can be named by imports are functions and wasm globals (not DATA symbols).

Not that the OptionalIndex is in the existing wasm function or global index space, not some new one.

This way the symbol table introduces exactly one new index space: the symbol index space.

(IsNamedByImport could also be wrapped in the flags).

Reply:
(In your idea, you'll need an index as well, in the case IsNamedByImport=false, to specify the function being exported!)

Another tweak to your idea: get rid of IsNamedByImport, simply have the Wasm index of the function/global. Then if it's an import, there's no name, and if it's not an import, a name follows.

I had a similar idea over the weekend. My refinement goes like this: You still have imports and exports. But, there's no need to refer to them by index. Imagine you have Fi function imports, Fe function exports. Then there must be at least Fi+Fe Function symbols: the first Fi refer to the imports (implicitly, no indexes written out), the next Fe symbols refer to the Fe function exports, and then any Function symbols on top of that contain a name+index to create a "hidden" export.